src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNIndexShard.java

```diff
@@ -0,0 +1,124 @@
+
+/**
+ * KNNIndexShard wraps IndexShard and adds methods to perform k-NN related operations against the shard
+ */
+public class KNNIndexShard {
+    private IndexShard indexShard;
+    private KNNIndexCache knnIndexCache;
+
+    private static Logger logger = LogManager.getLogger(KNNIndexShard.class);
+
+    /**
+     * Constructor to generate KNNIndexShard. We do not perform validation that the Index the IndexShard is from
+     * is in fact a k-NN Index (index.knn = true). This may make sense to add later, but for now the operations for
+     * KNNIndexShards that are not from a k-NN index should be no-ops.
+     *
+     * @param indexShard IndexShard to be wrapped.
+     */
+    public KNNIndexShard(IndexShard indexShard) {
+        this.indexShard = indexShard;
+        this.knnIndexCache = KNNIndexCache.getInstance();
+    }
+
+    /**
+     * Return the underlying IndexShard
+     *
+     * @return IndexShard
+     */
+    public IndexShard getIndexShard() {
+        return indexShard;
+    }
+
+    /**
+     * Return the name of the shards index
+     *
+     * @return Name of shard's index
+     */
+    public String getIndexName() {
+        return indexShard.shardId().getIndexName();
+    }
+
+    /**
+     * Load all of the HNSW graphs for this shard into the cache. Note that getIndices is called to prevent loading
+     * in duplicates.
+     *
+     * @return a List of KNNIndex's from this shard that are in the cache after this operation.
+     * @throws IOException Thrown when getting the HNSW Paths to be loaded in
+     */
+    public List<KNNIndex> warmup() throws IOException {
+        logger.info("[KNN] Warming up index: " + getIndexName());
+        Engine.Searcher searcher = indexShard.acquireSearcher("knn-warmup");
+        List<KNNIndex> indices;
+        try {
+            indices = knnIndexCache.getIndices(getHNSWPaths(searcher.getIndexReader()), getIndexName());
+            searcher.close();
+        } catch (IOException ex) {
+            searcher.close();
+            throw ex;
+        }
+        return indices;
+    }
+
+    /**
+     * For the given shard, get all of its HNSW paths
+     *
+     * @param indexReader IndexReader to read the file paths for the shard
+     * @return List of HNSW Paths
+     * @throws IOException Thrown when the SegmentReader is attempting to read the segments files
+     */
+    public List<String> getHNSWPaths(IndexReader indexReader) throws IOException {
+        List<String> hnswFiles = new ArrayList<>();
+        for (LeafReaderContext leafReaderContext : indexReader.leaves()) {
+            SegmentReader reader = (SegmentReader) FilterLeafReader.unwrap(leafReaderContext.reader());
+            hnswFiles.addAll(reader.getSegmentInfo().files().stream()
+                    .filter(fileName -> fileName.endsWith(getHNSWFileExtension(reader.getSegmentInfo().info)))
+                    .map(fileName -> shardPath().resolveIndex().resolve(fileName).toString())
```

Comment:
What if resolve(fileName) returns null?

Reply:
As part of the unit test for ```getHNSWPaths,  we could add a test to create a document in the knn index with out knn vector and refresh it and check the behavior. in this case the segment does not have hnsw files and we should assert it returns empty.

Reply:
Actually, fileName has the `@NotNull` annotation. So it cannot be null.

Reply:
Could you point me where annotation is?

Reply:
I see it in the source code in intellij. I think Intellij may have added the annotation.

![image](https://user-images.githubusercontent.com/19438237/87705964-938e2200-c753-11ea-8a8f-663c5190f418.png)

I am not sure what this implies. Do you know what this does @kaituo ?

Reply:
maybe https://www.jetbrains.com/help/idea/nullable-and-notnull-annotations.html#notnull